src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/TransportCreation/TcpTransportCreationHandler.cs

```diff
@@ -0,0 +1,193 @@
+        private const int DefaultPollTimeout = 30 * 1000000; // 30 seconds as microseconds
+        #endif
+
+        /// <inheritdoc />
+        public async ValueTask<Stream> Handle(ConnectionHandlerContext parameters, bool isAsync, CancellationToken ct)
+        {
+            // This handler cannot process if the protocol does not contain TCP
+            if (parameters.DataSource.ResolvedProtocol is not (DataSource.Protocol.Admin or DataSource.Protocol.TCP or DataSource.Protocol.None))
+            {
+                return null;
+            }
+
+            ct.ThrowIfCancellationRequested();
+
+            // DNS lookup
+            IPAddress[] ipAddresses = isAsync
+                ? await Dns.GetHostAddressesAsync(parameters.DataSource.ServerName, ct).ConfigureAwait(false)
+                : Dns.GetHostAddresses(parameters.DataSource.ServerName);
+            if (ipAddresses is null || ipAddresses.Length == 0)
+            {
+                throw new SocketException((int)SocketError.HostNotFound);
+            }
+
+            // If there is an IP version preference, apply it
+            switch (parameters.ConnectionString.IPAddressPreference)
+            {
+                case SqlConnectionIPAddressPreference.IPv4First:
+                    Array.Sort(ipAddresses, IpAddressVersionComparer.InstanceV4);
+                    break;
+
+                case SqlConnectionIPAddressPreference.IPv6First:
+                    Array.Sort(ipAddresses, IpAddressVersionComparer.InstanceV6);
+                    break;
+            }
+
+            // Attempt to connect to one of the matching IP addresses
+            // @TODO: Handle opening in parallel
+            Socket socket = null;
+            var socketOpenExceptions = new List<Exception>();
+
+            int portToUse = parameters.DataSource.ResolvedPort < 0
+                ? parameters.DataSource.Port
+                : parameters.DataSource.ResolvedPort;
+            var ipEndPoint = new IPEndPoint(IPAddress.None, portToUse); // Allocate once
+            foreach (IPAddress ipAddress in ipAddresses)
+            {
+                ipEndPoint.Address = ipAddress;
+                try
+                {
+                    socket = await OpenSocket(ipEndPoint, isAsync, ct).ConfigureAwait(false);
+                    break;
+                }
+                catch(Exception e)
+                {
+                    socketOpenExceptions.Add(e);
+                }
+            }
+
+            // If no socket succeeded, throw
+            if (socket is null)
+            {
+                // If there are any socket exceptions in the collected exceptions, throw the first
+                // one. If there are not, collect all exceptions and throw them as an aggregate.
+                foreach (Exception exception in socketOpenExceptions)
+                {
+                    if (exception is SocketException)
+                    {
+                        throw exception;
```

Comment:
We spoke offline about aligning with the way exceptions are surfaced from here to orchestrator. Please take a look at the error codes used by SNITcpHandle and how errors are created for them.
There are helpers now in SqlClientX added with the Prelogin PR which can help to go from a TCP provider to the SqlError and then that can be added to the error collection in the Context, which can then be used to create a SqlException.

Reply:
You will need to revisit all the exceptions being thrown out of here to align with what the current driver is throwing. The pattern is similar for all of them .